### PR TITLE
Framesize fix

### PR DESCRIPTION
**Pull request by kl0u**
In Apache Flink the results of the collect() call were returned through akka to the client. This led to an inherent limitation to the size of the output of a job, as this could not exceed the akka.framesize size. In other case, akka would drop the message.

To alleviate this, without dropping the benefits brought by akka and its out-of-the-box efficiency for small-sized results, we decided to keep forwarding the non-oversized (i.e. smaller than the akka.framesize) results through akka, and use the BlobCache module for the forwarding the oversized (large) ones.

Now the JobManager receives end merges the small accumulators (as before), and simply forwards to the Client the keys to the blobs storing the oversized ones. Now it is the responsibility of the Client to do the final merging between oversized and non-oversized accumulators.

### CODE REVIEW

**Comment by kl0u** [comment]
FLINK-2319 
This pull request targets this ticket.

**Comment by kl0u** [comment]
Hello guys,

This is a new pull request, for a previous ticket. 
It is aligned with recent changes in the master branch.

File: flink-runtime/src/main/java/org/apache/flink/runtime/accumulators/BaseAccumulatorSnapshot.java [hunk 1]
```diff
@@ -27,12 +27,7 @@
 import java.io.Serializable;
 import java.util.Map;
 
-/**
- * This class encapsulates a map of accumulators for a single task. It is used
- * for the transfer from TaskManagers to the JobManager and from the JobManager
- * to the Client.
- */
-public class AccumulatorSnapshot implements Serializable {
+public class BaseAccumulatorSnapshot implements Serializable {
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/accumulators/BaseAccumulatorSnapshot.java]: You removed the class description. Now you have to add a new one :)

File: flink-runtime/src/main/java/org/apache/flink/runtime/accumulators/LargeAccumulatorHelper.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+		try {
+			bc = new BlobClient(blobServerAddress);
+
+			for (Map.Entry<String, SerializedValue<Object>> entry : accumulators.entrySet()) {
+
+				String accumulatorName = entry.getKey();
+				byte[] accumulatorPayload = entry.getValue().getSerializedData();
+				if(accumulatorPayload == null) {
+					continue;
+				}
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/accumulators/LargeAccumulatorHelper.java]: Instead of using continue, could you handle the execution flow more explicit here?

File: flink-runtime/src/main/java/org/apache/flink/runtime/accumulators/BaseAccumulatorSnapshot.java [hunk 3]
```diff
@@ -44,18 +39,11 @@
...
-	private final SerializedValue<Map<String, Accumulator<?, ?>>> userAccumulators;
-
-	public AccumulatorSnapshot(JobID jobID, ExecutionAttemptID executionAttemptID,
-							Map<AccumulatorRegistry.Metric, Accumulator<?, ?>> flinkAccumulators,
-							Map<String, Accumulator<?, ?>> userAccumulators) throws IOException {
+	public BaseAccumulatorSnapshot(JobID jobID, ExecutionAttemptID executionAttemptID,
+			Map<AccumulatorRegistry.Metric, Accumulator<?, ?>> flinkAccumulators) throws IOException {
 		this.jobID = jobID;
 		this.executionAttemptID = executionAttemptID;
 		this.flinkAccumulators = new SerializedValue<Map<AccumulatorRegistry.Metric, Accumulator<?, ?>>>(flinkAccumulators);
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/accumulators/BaseAccumulatorSnapshot.java]: Why are the `flinkAccumulators` in `BaseAccumulatorSnapshot`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/accumulators/LargeAccumulatorSnapshot.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.api.common.accumulators.Accumulator;
+import org.apache.flink.runtime.blob.BlobKey;
+import org.apache.flink.runtime.executiongraph.ExecutionAttemptID;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+public class LargeAccumulatorSnapshot extends BaseAccumulatorSnapshot {
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/accumulators/LargeAccumulatorSnapshot.java]: Missing class description.

File: flink-runtime/src/main/java/org/apache/flink/runtime/accumulators/LargeAccumulatorSnapshot.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+	/**
+	 * In case some accumulators do not fit in akka, we store them in the blobCache and put
+	 * in the snapshot only the mapping between the name of the accumulator, and its blobKey
+	 * in the cache. This list holds exactly this mapping.
+	 * */
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/accumulators/LargeAccumulatorSnapshot.java]: > [..] do not fit in an Akka message payload [..]

File: flink-runtime/src/main/java/org/apache/flink/runtime/accumulators/LargeAccumulatorSnapshot.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+		this.largeUserAccumulatorBlobs = oversizedUserAccumulatorBlobKeys;
+	}
+
+	/**
+	 * Gets the Flink (internal) accumulators values.
+	 * @return the serialized map
+	 */
+	public Map<String, List<BlobKey>> getLargeAccumulatorBlobKeys() {
+		return largeUserAccumulatorBlobs;
+	}
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/accumulators/LargeAccumulatorSnapshot.java]: Is the java doc correct?

File: flink-runtime/src/main/java/org/apache/flink/runtime/accumulators/SmallAccumulatorSnapshot.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.api.common.accumulators.Accumulator;
+import org.apache.flink.runtime.executiongraph.ExecutionAttemptID;
+import org.apache.flink.runtime.util.SerializedValue;
+
+import java.io.IOException;
+import java.util.Map;
+
+public class SmallAccumulatorSnapshot extends BaseAccumulatorSnapshot {
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/accumulators/SmallAccumulatorSnapshot.java]: Class description.

File: flink-runtime/src/main/java/org/apache/flink/runtime/client/SerializedJobExecutionResult.java [hunk 8]
```diff
@@ -38,7 +42,13 @@
 
 	private final JobID jobId;
 
-	private final Map<String, SerializedValue<Object>> accumulatorResults;
+	// holds the result value of an accumulator
+	private final Map<String, SerializedValue<Object>> smallAccumulatorResults;
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/client/SerializedJobExecutionResult.java]: This should be `SerializedValue<Accumulator<?,?>>`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FallbackLibraryCacheManager.java [hunk 9]
```diff
@@ -33,6 +33,11 @@
 	private static Logger LOG = LoggerFactory.getLogger(FallbackLibraryCacheManager.class);
 
 	@Override
+	public int getBlobServerPort() {
+		return 0;
+	}
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FallbackLibraryCacheManager.java]: This is just a dummy return value. Do you check for it somewhere?

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java [hunk 10]
```diff
@@ -142,17 +146,25 @@
...
 			flinkAccumulators = accumulatorSnapshot.deserializeFlinkAccumulators();
-			userAccumulators = accumulatorSnapshot.deserializeUserAccumulators(userClassLoader);
+
+			if(accumulatorSnapshot instanceof SmallAccumulatorSnapshot) {
+				smallUserAccumulators = ((SmallAccumulatorSnapshot) accumulatorSnapshot).
+						deserializeSmallUserAccumulators(userClassLoader);
+			} else if(accumulatorSnapshot instanceof LargeAccumulatorSnapshot) {
+				largeUserAccumulators = ((LargeAccumulatorSnapshot) accumulatorSnapshot).
+						getLargeAccumulatorBlobKeys();
+			}
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java]: You could throw a RuntimeException otherwise.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java [hunk 11]
```diff
@@ -560,35 +572,118 @@ public ExecutionContext getExecutionContext() {
...
+	 * Gets a serialized map of the objects of the accumulators. This means that the actual
+	 * objects are serialized, thus merging can still be applied after deserialization.
+	 * @return The accumulator map with serialized accumulator objects.
+	 * @throws IOException
+	 */
+	public Map<String, SerializedValue<Object>> getSmallAccumulatorsSerialized() throws IOException {
+		return serializeAccumulators(false);
+	}
+
+	private Map<String,  SerializedValue<Object>> serializeAccumulators(boolean onlyContent) throws IOException {
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java]: Please document this method. Why did you add the `onlyContent` flag here?

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java [hunk 12]
```diff
@@ -599,14 +694,18 @@ public ExecutionContext getExecutionContext() {
 	 * @return an Array containing the StringifiedAccumulatorResult objects
 	 */
 	public StringifiedAccumulatorResult[] getAccumulatorResultsStringified() {
+		Map<String, Accumulator<?, ?>> smallAccumulatorMap = aggregateSmallUserAccumulators();
+		Map<String, List<BlobKey>> largeAccumulatorMap = aggregateLargeUserAccumulatorBlobKeys();
 
-		Map<String, Accumulator<?, ?>> accumulatorMap = aggregateUserAccumulators();
+		// get the total number of (unique) accumulators
+		Set<String> uniqAccumulators = new HashSet<String>();
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java]: unique

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java [hunk 13]
```diff
@@ -619,6 +718,21 @@ public ExecutionContext getExecutionContext() {
...
+				for (BlobKey bk : entry.getValue()) {
+					str.append(bk + " ");
+				}
+				str.append("]");
+
+				StringifiedAccumulatorResult result =
+						new StringifiedAccumulatorResult(entry.getKey(), "Blob/Serialized", str.toString());
+				resultStrings[i++] = result;
+			}
+		}
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java]: Shouldn't this be handled transparently for the user? The user shouldn't have to care if the accumulators were received from the BlobManager or via an Akka message.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java [hunk 14]
```diff
@@ -259,6 +261,10 @@ public Task(TaskDeploymentDescriptor tdd,
 		this.fileCache = checkNotNull(fileCache);
 		this.network = checkNotNull(networkEnvironment);
 
+
+		this.accumulatorRegistry = new AccumulatorRegistry(tdd.getJobConfiguration(),
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java]: space

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java [hunk 15]
```diff
@@ -309,6 +315,32 @@ public Task(TaskDeploymentDescriptor tdd,
...
+	 * */
+	private InetSocketAddress getBlobCacheServerAddress() {
+		if(jobManager == null || libraryCache == null) {
+			throw new RuntimeException("TaskManager not associated to JobManager.");
+		}
+
+		final String jmHost;
+		ActorRef jobManagerActor = this.jobManager.actor();
+		if (jobManagerActor == null) {
+			jmHost = "localhost";
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java]: "localhost" is hardcoded in several places.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java [hunk 16]
```diff
@@ -309,6 +315,32 @@ public Task(TaskDeploymentDescriptor tdd,
...
+		} else {
+			jmHost = jobManagerActor.path().address().host().getOrElse(
+					new AbstractFunction0<String>() {
+						@Override
+						public String apply() {
+							return "localhost";
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java]: Same here.

File: flink-runtime/src/main/java/org/apache/flink/runtime/util/SerializedValue.java [hunk 17]
```diff
@@ -47,6 +47,9 @@ public SerializedValue(T value) throws IOException {
 		this.serializedData = value == null ? null : InstantiationUtil.serializeObject(value);
 	}
 
+	public SerializedValue(byte[] data) {
+		this.serializedData = data == null ? null : data;
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/util/SerializedValue.java]: This could also be `this.serializedData = data`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/util/SerializedValue.java [hunk 18]
```diff
@@ -57,6 +60,25 @@ public T deserializeValue(ClassLoader loader) throws IOException, ClassNotFoundE
...
+		return serializedData;
+	}
+
+	/**
+	 * Gets the size of the serialized content.
+	 * @return the size in bytes of the data stored in the blob.
+	 * */
+	public long getSizeInBytes() {
+		if(serializedData == null) {
+			throw new RuntimeException("No data in this Blob.");
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/util/SerializedValue.java]: This Exception message is not general enough for the `SerializedValue` class.

File: flink-runtime/src/main/scala/org/apache/flink/runtime/akka/AkkaUtils.scala [hunk 19]
```diff
@@ -421,4 +421,16 @@ object AkkaUtils {
     val duration = Duration(ConfigConstants.DEFAULT_AKKA_LOOKUP_TIMEOUT)
     new FiniteDuration(duration.toMillis, TimeUnit.MILLISECONDS)
   }
+
+  def getFramesize(config: Configuration): Long = config.getString(
+    ConfigConstants.AKKA_FRAMESIZE,
+    ConfigConstants.DEFAULT_AKKA_FRAMESIZE).replaceAll("[^\\d.]", "").toLong
```
Comment (mxm) [inline flink-runtime/src/main/scala/org/apache/flink/runtime/akka/AkkaUtils.scala]: Why the "." in the regular expression?

File: flink-runtime/src/main/scala/org/apache/flink/runtime/akka/AkkaUtils.scala [hunk 20]
```diff
@@ -421,4 +421,16 @@ object AkkaUtils {
...
+
+  /**
+   * @return the threshold (in bytes) above which an object is considered too big
+   *         to transfer using akka. For now this parameter is set to 80% of the
+   *         akka.framesize.
+   * */
+  def getLargeAccumulatorThreshold(config: Configuration): Long =
+    (0.8 * getFramesize(config)).toLong
```
Comment (mxm) [inline flink-runtime/src/main/scala/org/apache/flink/runtime/akka/AkkaUtils.scala]: Maybe this should be configurable.

File: flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala [hunk 21]
```diff
@@ -338,16 +332,51 @@ class JobManager(
...
+                  smallAccumulatorResults.clear()
+
+                  // and update the blobKeys to send to the client.
+                  largeAccumulatorResults = executionGraph.
+                    addLargeUserAccumulatorBlobKeys(largeAccumulatorResults, newBlobKeys)
+
+                } else {
+                  // do nothing
+                  java.util.Collections.emptyMap()
+                }
```
Comment (mxm) [inline flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala]: The else branch can probably be removed? You could also let the two if branches return the value directly to the largeAccumulatorResults variable. Then it could be a `val` instead of a `var`.

File: flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala [hunk 22]
```diff
@@ -516,6 +545,19 @@ class JobManager(
...
 
   /**
+   * Gets the address where the blobCache is listening to.
+   * @return the address where the blobCache is listening to.
+   **/
+  private def getBlobCacheServerAddress: InetSocketAddress = {
+    if (libraryCacheManager == null) {
+      throw new RuntimeException("LibraryCacheManage is not initialized yet.")
+    }
+    val jmHost: String = "localhost"
```
Comment (mxm) [inline flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala]: "localhost" again. Maybe we can add a method to the BlobCache?

File: flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala [hunk 23]
```diff
@@ -911,6 +953,8 @@ class JobManager(
         try {
           eg.prepareForArchiving()
 
+          // todo KOSTAS: handle also the large accumulators.
+
```
Comment (mxm) [inline flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala]: TODO :)

File: flink-tests/src/test/java/org/apache/flink/test/misc/MiscellaneousIssuesITCase.java [hunk 24]
```diff
@@ -174,4 +177,52 @@ public void flatMap(Long value, Collector<Long> out) {
...
+			int noOfParallelism = 5;
+			int longsPerTask = 1200000;
+
+			env.setParallelism(noOfParallelism);
+			env.getConfig().disableSysoutLogging();
+
+			DataSet<Long> bigEnough = env.generateSequence(1, noOfParallelism * longsPerTask);
+			long theCount = bigEnough.collect().size();
+
+			assertEquals(noOfParallelism * longsPerTask, theCount);
```
Comment (mxm) [inline flink-tests/src/test/java/org/apache/flink/test/misc/MiscellaneousIssuesITCase.java]: You could also check the values of the oversized accumulators here.

File: flink-runtime/src/test/scala/org/apache/flink/runtime/testingUtils/TestingJobManager.scala [hunk 25]
```diff
@@ -181,7 +181,7 @@ trait TestingJobManager extends FlinkActor {
           currentJobs.get(jobID) match {
             case Some((graph, jobInfo)) =>
               val flinkAccumulators = graph.getFlinkAccumulators
-              val userAccumulators = graph.aggregateUserAccumulators
+              val userAccumulators = graph.aggregateSmallUserAccumulators
```
Comment (mxm) [inline flink-runtime/src/test/scala/org/apache/flink/runtime/testingUtils/TestingJobManager.scala]: This should also incorporate the blob cached accumulator results.

**Comment by mxm** [comment]
All in all, nice work @kl0u. Do you think we could change this pull request such that we can handle "small/large" or "message/blobcached" accumulators more transparently? It would be great if we encapsulated the two types within one message and only have one method `getUserAccumulators` to retrieve both types. The internal logic could be handled by either the `AccumulatorRegistry` or the `AccumulatorSnapshot` class. That would make the code easier to understand and remove some code from the JobManager and ExecutionGraph classes.

**Comment by kl0u** [comment]
Hi @mxm ,

Thanks a lot for the comments!

I integrated most of them. Please have a look and let me know what you think.

For the merging of the the different types of snapshots and handling them uniformly I do not have any current solution. If you have any, I am open, of course, to discuss it, because I agree that this would be nice.

For the comment on the getAccumulatorResultsStringified():
1) this is to be presented by the web interface to the user, just for monitoring purposes
2) this is called at the jobManager.

The problem is that the jobManager has only the blobKeys that point to the stored accumulators. The serialized data reside in the blobCache and have to be fetched in order to be inspected.

Currently the jobManager just forwards the blobKeys to the client, which fetches the blobs and does the deserialization and the final merging. This is done for jobManager scalability reasons, as given that we are talking about accumulators of arbitrary size, loading them from disk and deserializing them would be time and resource consuming. The same holds in the case that we wanted to get the type of these large accumulators (it is needed by the method). We would have to load and deserialize them at the jobManager. The currently implemented solution is just the result of this design decision. If you have any other strategy or solution that is worth implementing, let me know.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java [hunk 26]
```diff
@@ -75,24 +78,24 @@
...
+ * from the JobGraph's corresponding JobVertex.</li>
+ * <li>The {@link ExecutionVertex} represents one parallel subtask. For each ExecutionJobVertex, there are
+ * as many ExecutionVertices as the parallelism. The ExecutionVertex is identified by
+ * the ExecutionJobVertex and the number of the parallel subtask</li>
+ * <li>The {@link Execution} is one attempt to execute a ExecutionVertex. There may be multiple Executions
+ * for the ExecutionVertex, in case of a failure, or in the case where some data needs to be recomputed
+ * because it is no longer available when requested by later operations. An Execution is always
+ * identified by an {@link ExecutionAttemptID}. All messages between the JobManager and the TaskManager
+ * about deployment of tasks and updates in the task status always use the ExecutionAttemptID to
+ * address the message receiver.</li>
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java]: This change seems unnecessary.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java [hunk 27]
```diff
@@ -560,53 +621,177 @@ public ExecutionContext getExecutionContext() {
...
+//
+//		Map<String, SerializedValue<Object>> result = new HashMap<String, SerializedValue<Object>>();
+//		for (Map.Entry<String, Accumulator<?, ?>> entry : accumulatorMap.entrySet()) {
+//			Object toSerialize = onlyContent ? entry.getValue().getLocalValue() : entry.getValue();
+//			result.put(entry.getKey(), new SerializedValue<Object>(toSerialize));
+//		}
+//
+//		return result;
+//	}
+
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java]: Commented out.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java (hunk 13 above)
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java]: Ah ok. You're running into the same problem with the framesize here again.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java [hunk 28]
```diff
@@ -893,13 +1086,11 @@ public Object call() throws Exception {
...
 							break;
-						}
-						else if (numberOfRetriesLeft <= 0 && transitionState(current, JobStatus.FAILED, failureCause)) {
+						} else if (numberOfRetriesLeft <= 0 && transitionState(current, JobStatus.FAILED, failureCause)) {
 							postRunCleanup();
 							break;
 						}
-					}
-					else {
+					} else {
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java]: You're touching a lot of code here (and above) for no other reason than reformatting. We have no style guide yet that enforces this formatting. In general, we try not to touch existing lines to keep the latest history accurate.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java [hunk 29]
```diff
@@ -918,8 +1109,7 @@ private void postRunCleanup() {
 			if (coord != null) {
 				coord.shutdown();
 			}
-		}
-		catch (Exception e) {
+		} catch (Exception e) {
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java]: Same as above.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java [hunk 30]
```diff
@@ -966,8 +1162,7 @@ public boolean updateState(TaskExecutionState state) {
 					attempt.fail(new Exception("TaskManager sent illegal state update: " + state.getExecutionState()));
 					return false;
 			}
-		}
-		else {
+		} else {
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java]: Same as above.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java [hunk 31]
```diff
@@ -979,8 +1174,7 @@ public void scheduleOrUpdateConsumers(ResultPartitionID partitionId) {
 		if (execution == null) {
 			fail(new IllegalStateException("Cannot find execution for execution ID " +
 					partitionId.getPartitionId()));
-		}
-		else if (execution.getVertex() == null){
+		} else if (execution.getVertex() == null) {
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java]: Same as above.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java [hunk 32]
```diff
@@ -560,53 +621,177 @@ public ExecutionContext getExecutionContext() {
 	}
 
 	/**
+	 * This works as cache for already merged accumulators, as, in some cases,
+	 * we do not want to remerge accumulators as this may lead to duplicate entries.
+	 */
+	private Map<String, Accumulator<?, ?>> mergedSmallUserAccumulators;
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java]: Why can it lead to duplicate entries? The accumulators should always be merged from scratch. Every accumulator update message contains the current state of the accumulators.

**Comment by mxm** [comment]
Hi @kl0u ,

Thanks for addressing my comments. I've made some new ones :) but looks much better.

As for merging the different types of snapshots, the only solution I see now, is to let all the accumulator results go through the BlobCache. That might cause some extra latency at the end of the job. I'm not sure whether we should proceed on that path but the current solution has a lot of added overhead because we need to deal with two different kind of update mechanisms.

I would like to hear some other opinions on whether we should have both update mechanisms (BlobCache and Akka message-based) or just let everything go through the BlobCache. I'm leaning more towards the latter if it is not too slow for slow accumulators.

I think you have to rebase to the latest master again because of some minor changes.

**Comment by kl0u** [comment]
Hi @mxm ,

I totally agree that in terms of code clarity and structure, it would be better to let all accumulators pass through the BlobCache. In fact, we also had a brief discussion with @StephanEwen on the matter.

The only problem is the latency penalty that MAY result from this design choice. In other words, if for most of the usages akka is enough, then forcing small blobs to pass through the blobCache (i.e. reading and writing accumulators to disk) may be expensive.

What do you think?

**Comment by StephanEwen** [comment]
Agree. Especially with the frequently live updating accumulators, the small accumulators should stay in-frame

**Comment by kl0u** [comment]
Just rebased with the new version of the master.
Please have a look.

**Comment by StephanEwen** [comment]
I will have a look in a bit.

Am currently stuck with some streaming runtime fixes and the Kafka integration. Overflowing a bit with issues right now.

Thanks for bearing with us!

**Comment by kl0u** [comment]
No problem! 
This message was just a reminder.

Thanks a lot!

**Comment by mxm** [comment]
I see the point in keeping the Akka message-based transport of accumulators. @kl0u Sorry, I was also on vacations. Could you rebase to the current master again? Then, I think we can finally merge your changes.

**Comment by kl0u** [comment]
Hello!
I just rebased. Please have a look.

**Comment by mxm** [comment]
Hi @kl0u. I think something went wrong during rebasing.

**Comment by kl0u** [comment]
Hi @mxm. What do you mean?

**Comment by mxm** [comment]
From the GitHub UI it looks like your pull request branch and the Flink master branch have diverged, i.e. your changes are no longer based on a commit from the master branch.

**Comment by kl0u** [comment]
When I did the latest rebase it was saying that the two branches were ready to be merged.
Is there a way to see where they have diverged?

**Comment by mxm** [comment]
`git merge-base master framesize_fix` should show you the commit the branches diverged. Depending on the outcome, you might have to go back to an earlier version of your changes and rebase those against the master. It might be easier to squash all your commits before rebasing.

**Comment by kl0u** [comment]
Thanks @mxm. 
Although I don't think I will have time to fix it right now.

**Comment by mxm** [comment]
@kl0u That's fine. We'll be preparing a new milestone release based on the master. I think it makes sense to merge your changes afterwards for the 0.10 release.

**Comment by kl0u** [comment]
@mxm Sounds good.

**Comment by thorntree** [comment]
@kl0u ,hi kl0u.can you tell me how increase akka.framesize? I don't know how config akka.framesize.Thanks!

**Comment by TisonKun** [comment]
@thorntree https://ci.apache.org/projects/flink/flink-docs-stable/ops/config.html#akka-framesize

**Comment by thorntree** [comment]
@TisonKun I have set akka.framesize: "104857600b" in flink-conf.yaml,but it still doesn't work.

**Comment by TisonKun** [comment]
@thorntree what do you mean "it still doesn't work"? Could you provide more context such as log or output that you think so?

**Comment by TisonKun** [comment]
Besides, I think you should configure as

```yaml
akka.framesize: 104857600b
```

without the quote.
